rfc/0019-hashed-mojmap.md

```diff
@@ -0,0 +1,97 @@
+- The buildscript in Quilt's fork of Yarn will be updated to generate hashed Mojmap to Yarn mappings when built. When we start accepting PRs to our fork, the repository itself will be changed from using intermediary to hashed Mojmap. If necessary, a tool will be written to merge upstream from that point on.
+
+## Drawbacks
+
+- Hashed Mojmap names are more likely to change than intermediary, although not by as much as popularly believed.
+  - After applying the aforementioned corrections for package names, we found that Mojmap class names are twice as likely to change when the intermediary doesn't, than intermediary changes when Mojmap doesn't. 
+  - However, after manually checking the nature of these renames, it turned out that some of them were renamed to reflect a change in the purpose of the class, but the intermediary still matched because the classes were most similar. It is difficult to quantify exactly how common this is, as it requires going through all renamed classes manually.
+- It is possible that lambda methods change when intermediary doesn't, although this is not as likely in comparison to intermediary as popularly believed.
+  - Lambda methods are indeed included in Mojmap, and are in the javac format `lambda$enclosingMethod$index`.
+  - This name will only change if the enclosing method name changes, or lambdas are inserted before that lambda. Other lambdas being inserted into this method is also likely to fool Matcher and therefore change the intermediary.
+  - Not much effort is put into matching lambda methods that aren't auto-matched anyway. Lambda methods have been known to change intermediary with even the tiniest change in their content.
+- Mixed case names are harder to remember than all-lowercase names.
+  - Mixed case has been chosen to decrease the likelihood of a name collision, but it is still extremely unlikely that any two names will differ only by case.
+  - Any good mappings lookup should be case-insensitive.
+  - Eclipse is known to have case-sensitive auto-completion, but the right class/field/method will come up after typing only the first couple of characters, making it unnecessary to remember the whole thing for auto-completion.
+
+## Rationale and Alternatives
+
+- We could continue using Fabric intermediary.
+  - The problems mentioned in the "Motivation" section would then persist.
+  - We would miss our opportunity to change from intermediary without breaking mods, and would have to break mods if we decide to change in the future.
+- We could fork intermediary and diverge but otherwise keep it the same.
+  - Does not improve the maintenance burden of updating to the latest Minecraft version.
+  - Does not handle forks of Minecraft versions (e.g. April fools snapshots) very well.
+- We could auto-match using Mojmap using the aforementioned techniques to minimize the chance of bad renames.
```

Comment:
We do have prior art regarding matching only Mojang names (albeit with some lambda matching help), see here <https://github.com/MinecraftForge/DePigifier>